File: DSA/AnshulJain_590014934/Day_37_Question_2.py
def evaluate_postfix(expression):
    stack = []
    for token in expression:
        if token.isdigit():
            stack.append(int(token))
        elif len(stack) >= 2:
            b = stack.pop()
            a = stack.pop()
            if token == '+':
                stack.append(a + b)
            elif token == '-':
                stack.append(a - b)
            elif token == '*':
                stack.append(a * b)
            elif token == '/':
                stack.append(int(a / b))
    return stack[0]
expression = input("Enter a postfix expression: ") #taking input from user
result = evaluate_postfix(expression)
print(result)
#for example:
#Input:231*+9-
#Output:-4

File: DSA/Animesh_jha_590016813/Day_37_question_2.c
#include <stdio.h>
#include <stdlib.h>
#include <ctype.h>

struct stack {
    int size;
    int top;
    float *arr;
};

int isEmpty(struct stack *ptr) {
    return ptr->top == -1;
}

void push(struct stack *ptr, float val) {
    ptr->top++;
    ptr->arr[ptr->top] = val;
}

float pop(struct stack *ptr) {
    if (isEmpty(ptr)) {
        printf("Stack underflow! Cannot pop.\n");
        return -1;
    } else {
        float val = ptr->arr[ptr->top];
        ptr->top--;
        return val;
    }
}

float evaluatePostfix(char *postfix) {
    struct stack *sp = (struct stack *)malloc(sizeof(struct stack));
    sp->size = 100;
    sp->top = -1;
    sp->arr = (float *)malloc(sp->size * sizeof(float));

    for (int i = 0; postfix[i] != '\0'; i++) {
        if (isdigit(postfix[i])) {
            push(sp, postfix[i] - '0');  // Convert char to int
        } else {
            float operand2 = pop(sp);
            float operand1 = pop(sp);
            switch (postfix[i]) {
                case '+': push(sp, operand1 + operand2); break;
                case '-': push(sp, operand1 - operand2); break;
                case '*': push(sp, operand1 * operand2); break;
                case '/': push(sp, operand1 / operand2); break;
            }
        }
    }

    float result = pop(sp);
    free(sp->arr);
    free(sp);
    return result;
}

int main() {
    char postfix[] = "231*+9-";
    printf("Result of postfix evaluation: %.2f\n", evaluatePostfix(postfix));
    return 0;
}
